### ISSUE TITLE
Batch commands functionality

### ISSUE THREAD

**Issue by simontabor** [NodeRedis/node_redis#806]
Related to discussions in https://github.com/NodeRedis/node_redis/pull/799

It's very common to want to run a batch of commands at the same time and have a single callback for when they are completed. A lot of the time, people use `MULTI` for this, which is okay but adds noticeable overhead when pushing Redis hard.

An example of this is at GoSquared, where we add a new visitor to our tracking. This involves many redis commands - setting a hash of the visitor details and adding them to relevant lists what we want to run at the same time, but we don't need a multi.

As mentioned in the previous issue, I've got the functionality in redis clustr (https://github.com/gosquared/redis-clustr/blob/master/src/RedisBatch.js) and it's been incredibly useful. Usage is as follows:

```js
var batch = redis.batch();
batch.set('key', 'value');
batch.get('key', function(err, resp) {
  // "value"
});;
batch.exec(function(err, resp) {
  // [ "OK", "value" ]
});
```

Trying to use something like `async` to group the commands together results in awful code, this is familiar and clear (exactly the same usage as multi).

Thoughts?

**Comment by stockholmux** [comment]
So, the above isn't atomic like running `multi`? 

I'm interested in an optimized `multi`-like block that will efficiently run a single command if that is all that is needed. Here is my use case:

```
var myAtomicBatch = client.atomicBatch();

myAtomicBatch.set('key','value');
if (isSomething === true) {
  myAtomicBatch.set('key-for-is-something','value);
}
myAtomicBatch.exec(function(err, values) {
  //...
});
```

So the second `set` command is only called under certain condition(s) but I need both the first and second commands to run together if it is called. 

We do something similar to this now but just take the performance hit if the conditional is not `true`. If I wanted to refactor for to avoid the multi / single command issue, the code would come out fairly messy.

**Comment by simontabor** [comment]
Exactly - this will handle that for you

**Comment by stockholmux** [comment]
@simontabor Gotcha - the first callback and the part about using `async` made me think you were going for a non-atomic series of commands. Also, the link to RedisBatch.js is 404'ing for me.

**Comment by simontabor** [comment]
@stockholmux edited the link, for some reason it had a full stop at the end.

Think I've misunderstood slightly - you're looking for it to run a multi/exec block if it's more than one command but just run the single command if there's only one?

This will simply throw all the commands at redis when you call exec (non-atomic). I understand the use case for an atomic batch - that could quite easily be added too. I've not actually experienced any issues with using the non-atomic batching, especially as we use Redis Cluster and that makes multi's useless anyway.

**Comment by stockholmux** [comment]
@simontabor Right - more of an optimization for `multi` I suppose. 
```
var myAtomicBatch = client.atomicBatch();
myAtomicBatch.set('key','value');
myAtomicBatch.exec(function(err, values) {
  //...
});
```
is optimized (internally) to something like
```
client.set('key','value', function(err, value) {
  if (err) { cb(err); } else { cb(err,[values]); } 
});
```

But, if the multi queue is greater than 1, then the multi is called as normal.

It could be shimmed at the application level, but it seems like a simple / smart optimization that could make a fair bit of difference at scale.

**Comment by simontabor** [comment]
Definitely +1, that's what I did in https://github.com/NodeRedis/node_redis/pull/799.

I see value in both an atomic and a non-atomic version - WDYT @stockholmux?

**Comment by stockholmux** [comment]
@simontabor If the syntax remains more-or-less the same for optimized multi, 'normal' multi and non-atomic batch, then the end-user of the module could decide what is best without changing much code. That could be a big win.

**Comment by BridgeAR** [comment]
This is implemented by #878 and will be released in v.2.2.0.
